### PR TITLE
[SPARK-23360][SQL][PYTHON] Get local timezone from environment via pytz, or dateutil.

### PR DESCRIPTION
**Pull request by ueshin**
## What changes were proposed in this pull request?

Currently we use `tzlocal()` to get Python local timezone, but it sometimes causes unexpected behavior.
I changed the way to get Python local timezone to use pytz if the timezone is specified in environment variable, or timezone file via dateutil .

## How was this patch tested?

Added a test and existing tests.

### CODE REVIEW

**Comment by ueshin** [comment]
cc @icexelloss @BryanCutler @felixcheung @HyukjinKwon @cloud-fan

File: python/pyspark/sql/types.py [hunk 1]
```diff
@@ -1709,6 +1709,15 @@ def _check_dataframe_convert_date(pdf, schema):
     return pdf
 
 
+def _get_local_timezone():
+    """ Get local timezone from environment vi pytz, or dateutil. """
+    from pyspark.sql.utils import require_minimum_pandas_version
+    require_minimum_pandas_version()
+
+    import os
+    return os.environ.get('TZ', 'dateutil/:')
```
Comment (icexelloss) [inline python/pyspark/sql/types.py]: I don't really understand how does "dateutil/:" work, can you maybe add some comments for that?

**Comment by icexelloss** [comment]
Thanks @ueshin for the quick patch!

File: python/pyspark/sql/tests.py [hunk 2]
```diff
@@ -4124,7 +4126,7 @@ def test_vectorized_udf_timestamps(self):
         data = [(0, datetime(1969, 1, 1, 1, 1, 1)),
                 (1, datetime(2012, 2, 2, 2, 2, 2)),
                 (2, None),
-                (3, datetime(2100, 3, 3, 3, 3, 3))]
+                (3, datetime(2100, 4, 4, 4, 4, 4))]
```
Comment (BryanCutler) [inline python/pyspark/sql/tests.py]: Just wondering if changing these values made a difference somewhere?

File: python/pyspark/sql/types.py [hunk 3]
```diff
@@ -1709,6 +1709,15 @@ def _check_dataframe_convert_date(pdf, schema):
     return pdf
 
 
+def _get_local_timezone():
+    """ Get local timezone from environment vi pytz, or dateutil. """
```
Comment (BryanCutler) [inline python/pyspark/sql/types.py]: did you mean "from environment via pytz"?  the 'TZ' environment var is read by the `datetime` module, does pytz do anything with this?

**Review by BryanCutler — COMMENTED** [review]
Thanks @ueshin for the fix! Do we know why `tzlocal()` was inconsistent - is it because of DST and ambiguous flag?

File: python/pyspark/sql/types.py [hunk 4]
```diff
@@ -1709,6 +1709,15 @@ def _check_dataframe_convert_date(pdf, schema):
...
+    from pyspark.sql.utils import require_minimum_pandas_version
+    require_minimum_pandas_version()
```
[the remark this answers is not in the document]
Comment (ueshin) [reply python/pyspark/sql/types.py]: Actually, it isn't needed here. I'll remove it. Thanks.

File: python/pyspark/sql/types.py (hunk 1 above)
Comment (ueshin) [reply python/pyspark/sql/types.py]: Sure, I'll add some comments.

File: python/pyspark/sql/tests.py (hunk 2 above)
Comment (ueshin) [reply python/pyspark/sql/tests.py]: I'll revert it.

File: python/pyspark/sql/types.py (hunk 3 above)
Comment (ueshin) [reply python/pyspark/sql/types.py]: I modified and added the comment. Can you understand by the comment?

**Comment by ueshin** [comment]
@BryanCutler Seems like pandas handles `tzlocal()` differently than other timezone, and it might handle DST incorrectly, I guess.

File: python/pyspark/sql/types.py [hunk 5]
```diff
@@ -1766,15 +1781,13 @@ def _check_series_convert_timestamps_localize(s, from_timezone, to_timezone):
...
+    from_tz = from_timezone or _get_local_timezone()
+    to_tz = to_timezone or _get_local_timezone()
     # TODO: handle nested timestamps, such as ArrayType(TimestampType())?
     if is_datetime64tz_dtype(s.dtype):
         return s.dt.tz_convert(to_tz).dt.tz_localize(None)
     elif is_datetime64_dtype(s.dtype) and from_tz != to_tz:
-        # `s.dt.tz_localize('tzlocal()')` doesn't work properly when including NaT.
-        return s.apply(lambda ts: ts.tz_localize(from_tz).tz_convert(to_tz).tz_localize(None)
-                       if ts is not pd.NaT else pd.NaT)
+        return s.dt.tz_localize(from_tz).dt.tz_convert(to_tz).dt.tz_localize(None)
```
Comment (HyukjinKwon) [inline python/pyspark/sql/types.py]: @ueshin, is it safe to remove `if ts is not pd.NaT else pd.NaT`? Seems there is a small possibility for `tzlocal()`:

https://github.com/pandas-dev/pandas/blob/0.19.x/pandas/tslib.pyx#L1760
https://github.com/pandas-dev/pandas/blob/0.19.x/pandas/tslib.pyx#L54
https://github.com/dateutil/dateutil/blob/2.6.1/dateutil/tz/tz.py#L1362
https://github.com/dateutil/dateutil/blob/2.6.1/dateutil/tz/tz.py#L1408

File: python/pyspark/sql/tests.py [hunk 6]
```diff
@@ -2867,6 +2867,35 @@ def test_create_dataframe_required_pandas_not_found(self):
...
+        orig_env_tz = os.environ.get('TZ', None)
+        orig_session_tz = self.spark.conf.get('spark.sql.session.timeZone')
+        try:
+            tz = 'America/Los_Angeles'
+            os.environ['TZ'] = tz
+            time.tzset()
+            self.spark.conf.set('spark.sql.session.timeZone', tz)
+
+            df = self.spark.createDataFrame(pdf)
+            df.show()
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests.py]: gentle reminder for it. Seems it's now there for debugging purpose I guess? :).

**Review by HyukjinKwon — APPROVED** [review]
LGTM except a question and a nit.

File: python/pyspark/sql/types.py (hunk 5 above)
Comment (ueshin) [reply python/pyspark/sql/types.py]: Good catch! I'll revert this.

File: python/pyspark/sql/tests.py (hunk 6 above)
Comment (ueshin) [reply python/pyspark/sql/tests.py]: Oops, I should've removed it. Thanks!

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Merged to master and branch-2.3.

**Pull request closed without merging**